File: src/Scripting/Ruby/patterns/publish_subscribe.rb
# frozen_string_literal: true

module RubyPatterns
  module PublishSubscribe
    module_function

    def run
      topics = Hash.new { |hash, key| hash[key] = [] }
      received = []
      topics[:news] << ->(value) { received << value }
      topics[:news].each { |subscriber| subscriber.call('v1') }
      raise 'Publish-Subscribe failed' unless received == ['v1']
      true
    end
  end
end

RubyPatterns::PublishSubscribe.run if $PROGRAM_NAME == __FILE__

File: src/Scripting/Lua/patterns/publish_subscribe.lua
local topics = {}
local function subscribe(topic, fn)
  topics[topic] = topics[topic] or {}
  table.insert(topics[topic], fn)
end
local function publish(topic, value)
  for _, fn in ipairs(topics[topic] or {}) do fn(value) end
end
local received = {}
subscribe("news", function(v) table.insert(received, v) end)
publish("news", "release")
assert(received[1] == "release")
return true
